File: Codility/07. Stacks and Queues/Brackets/Solution.swift
// https://app.codility.com/programmers/lessons/7-stacks_and_queues/brackets/
// Painless

public func solution(_ S : inout String) -> Int {
    func isMatchingPair(_ A : Character, _ B : Character) -> Bool {
        return (A == "(" && B == ")") || (A == "[" && B == "]") || (A == "{" && B == "}")
    }

    var stack = [Character]()

    for c in S {
        if c == "(" || c == "[" || c == "{" {
            stack.append(c)
        } else {
            if stack.isEmpty {
                return 0
            }

            if !isMatchingPair(stack.popLast()!, c) {
                return 0
            }
        }
    }

    return stack.isEmpty ? 1 : 0
}

File: Codility/07. Stacks and Queues/Brackets/Solution.py
# https://app.codility.com/programmers/lessons/7-stacks_and_queues/brackets/
# Painless


def solution(S: str) -> int:
    """Push openers on a stack and pop on closers, requiring a matching opener for balance."""
    # Time: O(n)   Space: O(n)
    pairs = {")": "(", "]": "[", "}": "{"}
    stack: list[str] = []

    for c in S:
        if c in "([{":
            stack.append(c)
        else:
            if not stack or stack.pop() != pairs[c]:
                return 0

    return 1 if not stack else 0


def test() -> None:
    assert solution("{[()()]}") == 1
    assert solution("([)()]") == 0
    assert solution("") == 1
    assert solution("(") == 0
    assert solution(")(") == 0
